# InbreedingDepressionSeychellesWarbler
 
These are the data and scripts used to investigate individual inbreeding depression using 1.936 low coverage, whole-genome Seychelles warbler sequences that have been imputed to recover missing genotypes. The pipeline to impute SNPs and verify accuracy can be found in my other repository: https://github.com/kiran-lee/SNPsSeychellesWarbler/tree/main and the imputed SNPs can be found as a compressed vcf file on Zenodo: https://zenodo.org/records/12570527. 

An explanation of the files in Data can be found in the InbreedingDepressionSeychellesWarblers r markdown script.

Perhaps the most useful file to other researchers is a dataframe (SeychellesWarblerTraitsCorrected.txt) of all birds in the Seychelles warbler database (BirdID), the plate number they were sequenced on (Plate), sequences sample name as named by Liverpool University (SeqID),  sequencing coverage (Coverage), tube number of sample used to link SeqID to BirdID  (ID), whether the ID used was BloodID or BloodTubeNumber (Identifier) FROH inbreeding coefficient from recent inbreeding after bottleneck (LargeFROHbin),  inbreeding around bottleneck (MediumFROHbin), and inbreeding before bottleneck (SmallFROHbin), lifespan (Lifespan), year of birth (BirthYear), year it was last seen (LastSeenYear), lifetime offspring produced (ReproductiveOutput), mean death rate between field seasons an individual has lived through as an environmental control of lifetime hardship faced (MeanLifeTimeDeathRate), 
mother ID (DamID), father ID (SireID).